app/javascript/packs/referral_charts.jsx

```diff
@@ -0,0 +1,183 @@
+            fill: false,
+          },
+          {
+            label: 'Installs',
+            data: data.map(x => x.first_runs),
+            borderColor: '#66C3FC',
+            fill: false,
+          },
+          {
+            label: '30-Day-Use',
+            data: data.map(x => x.finalized),
+            borderColor: '#7B82E1',
+            fill: false,
+          },
+        ]
+      },
+      options: {
+        tooltips: {
+          mode: 'x'
+        },
+        title: {
+          fontSize: 18,
+          display: true,
+          text: title.toUpperCase()
+        },
+        scales: {
+          xAxes: [{
+            type: 'time',
+            distribution: 'series',
+            ticks: {
+              source: 'labels'
+            },
+            time: {
+              unit: 'day'
+            }
+          }],
+          yAxes: [{
+            type: 'logarithmic',
+            ticks: {
+              suggestedMax: suggestedMax
+            }
+          }]
+        },
+        legend: {
+          labels: {
+            usePointStyle: true
+          }
+        }
+      }
+    });
+  }
+
+  async viewReferralCodeStats() {
+    const node = this.selectMenuRef.current;
+    var url = routes.publishers.promo_registrations.show.path.replace('{id}', document.getElementById('publisher_id').value);
+    url = url.replace('{referral_code}', node.state.value);
+    const result = await fetch(url, {
+      headers: {'Accept': 'text/html', 'X-Requested-With': 'XMLHttpRequest', 'X-CSRF-Token': document.head.querySelector("[name=csrf-token]").content},
+      credentials: 'same-origin',
+      method: "GET"
+    }).then(response => {
+      response.json().then(json => {
+        if (json !== undefined && json.length != 0) {
+          this.createChart(json, node.state.value, this.getSuggestedMax(json));
+        }
+      });
+    });
+  }
+
+  render() {
+    var referralCodesForSelect = [];
+    this.state.referralCodes.forEach(function(element) {
+      referralCodesForSelect.push(<div data-value={element}>{element}</div>);
+    });
+    return (
+      <div style={{display: 'inline-flex', flexDirection: 'row', justifyContent: 'flex-start'}}>
+        <ControlWrapper
+          text={'Choose a Referral Code to view its Stats (ones with no records are hidden)'}
+          type={'light'}>
+            <div style={{maxWidth: "350px"}} >
+              <Select type={'light'} ref={this.selectMenuRef} >
+                {referralCodesForSelect}
+              </Select>
+            </div>
+        </ControlWrapper>
+        <div>
+          <div style={{marginTop: '15px', marginLeft: '15px'}}>
+            <PrimaryButton onClick={this.viewReferralCodeStats} enabled={true} >
+              View
+            </PrimaryButton>
+          </div>
+        </div>
+      </div>
+    );
+  }
+}
+
+export function renderReferralCharts() {
+  let props = {
+    referralCodes: JSON.parse(document.getElementById('referrals-hidden-tags').value)
```

Comment:
might be better to do something like this

const { value } = document.getElementById('referrals-hidden-tags')
if(value === undefined) {
  return;
}
let referralCodes = JSON.parse(value);